torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -398,19 +411,63 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
       : bdimx + (device_warp_size - bdimx % device_warp_size);
 
   bool pad_bdimx = bdimx > 16 &&
-      padded_bdimx * bdimy * bdimz <
-          (int64_t)at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;
+      padded_bdimx * bdimy * bdimz < (int64_t)dev_prop->maxThreadsPerBlock;
 
   pad_bdimx = pad_bdimx &&
       bdimx * inner_reduction_unroll_factor != inner_most_dimension_numel;
 
+  int64_t nvrtc_register_per_thread = LaunchParams::UNINITIALIZED_VAL;
+  const int blocksPerKernel = godim;
+  if (vectorize && blocksPerKernel > device_multiprocessor_count &&
+      batches_per_block_inner_reduction > 1) {
+    // estimate register usage and occupancy raito.
+    // If occupancy raito is less than occupancy_ratio, reduce register usage
+    constexpr int reg_allocation_granularity = 256;
+    constexpr float occupancy_ratio = 0.4;
+    const int persistent_buffer_size = batches_per_block_inner_reduction *
+        inner_reduction_unroll_factor * max_input_dtype_size;
+    // persistent_buffer_size = 4*2, 8*2, 32*2, 64*2, 128*2
+    // register_used_on_a100  = 27,  40,  62,   73,   105
+    // register_used_on_v100  = xx,  xx,  45,   62,   93
+    // estimated_register_num = 42,  44,  56,   72,   104
+    // safe for both v100 & a100
+    const int estimated_register_count = persistent_buffer_size / 4 + 40;
+    const int register_per_warp =
+        ceilDiv(
+            estimated_register_count * device_warp_size,
+            reg_allocation_granularity) *
+        reg_allocation_granularity;
+    const int threadsPerBlock =
+        (pad_bdimx ? padded_bdimx : bdimx) * bdimy * bdimz;
+    const int warps_per_block = ceilDiv(threadsPerBlock, dev_prop->warpSize);
+    const int estimated_warps_per_sm = dev_prop->regsPerMultiprocessor /
+        (register_per_warp * warps_per_block) * warps_per_block;
+    const int occupancy_warps_per_sm = static_cast<int>(
+        dev_prop->maxThreadsPerMultiProcessor / device_warp_size *
+        occupancy_ratio);
+    if (estimated_warps_per_sm < occupancy_warps_per_sm) {
+      const int blocks_per_sm_1 = dev_prop->maxBlocksPerMultiProcessor;
+      const int blocks_per_sm_2 =
+          ceilDiv(occupancy_warps_per_sm, warps_per_block);
+      const int blocks_per_sm = std::min(blocks_per_sm_1, blocks_per_sm_2);
+      const int warps_per_sm = blocks_per_sm * warps_per_block;
+      const int register_per_warp = dev_prop->regsPerMultiprocessor /
+          warps_per_sm / reg_allocation_granularity *
+          reg_allocation_granularity;
+      const int occupancy_register_count = register_per_warp / device_warp_size;
+      nvrtc_register_per_thread = std::max(
```

Comment:
I don't follow why `max` is used here. It seems that this part is only executed when the estimated occupancy based on the estimated register count is lower than the target occupancy of 0.4, so shouldn't `occupancy_register_count` be smaller than `estimated_register_count` and we'd need to use `occupancy_register_count` to make the occupancy as high as the target?

Reply:
I'm actually unclear if this occupancy target should be considered or should be considered here. If it should be considered, I think we should adjust the buffer size. For example, at line 278, `warps_per_block` is recursively halved with the minimum of 8. That minimum number should be determined by the target occupancy ratio, shouldn't it?

Reply:
1) `max` is used to avoid the potential register spill due to the aggressive adjust of registers.
2) if we want to estimate `warps_per_block` using `occupancy_ratio`, we need to know `estimated_register_count`, which depends on `persistent_buffer_size`, which depends on `warps_per_block`. It is cyclic.

Reply:
Isn't using `max` negating the intent of `estimated_warps_per_sm < occupancy_warps_per_sm`? We first estimate the occupancy just using the estimated register count, and then if that's lower than the target occupancy, we also calculate the maximum available register count that can achieve the occupancy. That latter register count shouldn't be smaller than the former count? Using `max` here seems to mean we just ignore the latter register limit.

Reply:
the estimated_warps_per_sm is scaled by a factor less than 1. so we are not ignore the latter register limit.

Reply:
Oh, so if the occupancy-based estimate is smaller than the persistent-buffer-based estimate, we use the former but we cap the decrease within 20% of the latter. Am I understanding it correctly?

Reply:
Please leave more detailed comments at the top of the if block about the reasoning.

Reply:
Yes. we allow a maximum of 20% discount on `estimated_register_count` to make occupancy estimation happy, although it may still be higher than what occupancy estimation asked for.